app/api/chat/agent_hil/route.ts

```diff
@@ -0,0 +1,135 @@
+import { NextRequest, NextResponse } from "next/server";
+import { Message as VercelChatMessage, StreamingTextResponse } from "ai";
+
+import {
+  AIMessage,
+  BaseMessage,
+  ChatMessage,
+  HumanMessage,
+} from "@langchain/core/messages";
+import { createAgent } from "./agent_creator";
+import { Command } from "@langchain/langgraph";
+
+const convertVercelMessageToLangChainMessage = (message: VercelChatMessage) => {
+  if (message.role === "user") {
+    return new HumanMessage(message.content);
+  } else if (message.role === "assistant") {
+    return new AIMessage(message.content);
+  } else {
+    return new ChatMessage(message.content, message.role);
+  }
+};
+
+const convertLangChainMessageToVercelMessage = (message: BaseMessage) => {
+  if (message._getType() === "human") {
+    return { content: message.content, role: "user" };
+  } else if (message._getType() === "ai") {
+    return {
+      content: message.content,
+      role: "assistant",
+      tool_calls: (message as AIMessage).tool_calls,
+    };
+  } else {
+    return { content: message.content, role: message._getType() };
+  }
+};
+
+export async function POST(req: NextRequest) {
+  try {
+    const body = await req.json();
+    const returnIntermediateSteps = body.show_intermediate_steps;
+    /**
+     * We represent intermediate steps as system messages for display purposes,
+     * but don't want them in the chat history.
+     */
+    const messages = (body.messages ?? [])
+      .filter(
+        (message: VercelChatMessage) =>
+          message.role === "user" || message.role === "assistant",
+      )
+      .map(convertVercelMessageToLangChainMessage);
+
+    // Setup checkpointer DO THIS ONCE if you haven't.
+    // await setupCheckpointer();
+
+    const agent = createAgent();
+
+    const threadId = "1999"; // TODO: get from request
+
+    // Get the agent state, check whether the agent can be resumed.
+    const state = await agent.getState({
+      configurable: { thread_id: threadId },
+    });
+
+    const isBeginning =
+      state.next.length === 0 && Object.keys(state.values).length === 0;
+
+    let runInput: { messages: (AIMessage | HumanMessage)[] } | Command =
+      new Command({
+        resume: messages.findLast((m: any) => m.role === "user")?.content,
+        update: {
+          isResuming: true, // Important if you have workflows that you want to ignore when resuming
+        },
+      });
+
+    if (isBeginning) {
+      runInput = {
+        messages,
+      };
+    }
```

Comment:
there's gotta be a simpler way to do this, I simply want to check whether the current state is resumable or not, or whether it is in its __end__ node?

I'm hoping for something like `state.currentNode` to get the string name of the node, if it is `__end__` then I know it's not resumable. something like this exist?

Reply:
Wouldn't it just be if `state.next` were populated?